### ISSUE TITLE
"Undo" inserts extra ^M if file has CRLF line endings on Linux

### ISSUE THREAD

**Issue by artem-nefedov** [airblade/vim-gitgutter#657]
On Linux, if the file has CRLF line endings, calling `<leader>hu` on modified chunk inserts extra `^M` at the end of each line, and still shows line as modified (`~`) afterwards, until `^M` is deleted manually. Subsequent undo calls do not help.

> What is the latest commit SHA in your installed vim-gitgutter?

1e5d8c8f142f5f0c4ede3c04322ae969b12ede30

> What vim/nvim version are you on?

```
:version
NVIM v0.3.8
Build type: Release
LuaJIT 2.0.4
Compilation: /usr/bin/x86_64-linux-gnu-gcc -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIF
Y_SOURCE=1 -DDISABLE_LOG -Wdate-time -D_FORTIFY_SOURCE=1 -Wconversion -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unuse
d-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATE
D_DECLARATIONS -D_GNU_SOURCE -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim-4YEfmh/neovim-0.3.8/build/config -I/build/neovim-4YEfmh/neovim-0
.3.8/src -I/usr/include -I/build/neovim-4YEfmh/neovim-0.3.8/build/src/nvim/auto -I/build/neovim-4YEfmh/neovim-0.3.8/build/include
Compiled by team+vim@tracker.debian.org

Features: +acl +iconv -jemalloc +tui
```

**Comment by airblade** [comment]
Are you saying that when you undo a modified chunk (with `<leader>hu`) you get CRLF^M line endings?

**Comment by artem-nefedov** [comment]
The other way around, `^MCRLF`, so actually just `CRCRLF`.

**Comment by airblade** [comment]
Ok.  What line-ending config does the repo have?

**Comment by artem-nefedov** [comment]
There are no specific settings defined, so it's on per-file basis.
I assume that, since I'm on Linux, it expects `LF` by default.
But, I never had any problems with internal tools.
For file with `CRLF` line endings, output of `git --no-pager diff` looks normal:
```
-Original line
+Modified line
```

**Comment by airblade** [comment]
> I assume that, since I'm on Linux, it expects LF by default.

No, it tries to get line endings right:

https://github.com/airblade/vim-gitgutter/blob/1725c13add66c6981d9406f109dcd64dff5fbf59/autoload/gitgutter/diff.vim#L37-L61

If gitgutter shows the signs you expect when you modify lines in the file, it would suggest the problem lies in undo part of the code rather than the diffing.

**Comment by airblade** [comment]
Does this solve the problem for you?

```diff
diff --git i/autoload/gitgutter/hunk.vim w/autoload/gitgutter/hunk.vim
index ffde78b..ea8db03 100644
--- i/autoload/gitgutter/hunk.vim
+++ w/autoload/gitgutter/hunk.vim
@@ -278,7 +278,7 @@ endfunction
 function! s:undo(hunk_diff)
   " Apply reverse patch to buffer.
   let hunk  = gitgutter#diff#parse_hunk(split(a:hunk_diff, '\n')[4])
-  let lines = map(split(a:hunk_diff, '\n')[5:], 'v:val[1:]')
+  let lines = map(split(a:hunk_diff, '\r\?\n')[5:], 'v:val[1:]')
   let lnum  = hunk[2]
   let added_only   = hunk[1] == 0 && hunk[3]  > 0
   let removed_only = hunk[1]  > 0 && hunk[3] == 0
```

**Comment by artem-nefedov** [comment]
Just tested it. This indeed removes the issue.
Please, commit to upstream.
(hope this doesn't bring side effects on Windows)

**Comment by airblade** [comment]
Fixed in 320b7a3.  Thanks for reporting it.
